src/Compilers/CSharp/Portable/Symbols/Source/SourceTypeParameterSymbol.cs

```diff
@@ -28,6 +28,17 @@ internal abstract class SourceTypeParameterSymbolBase : TypeParameterSymbol, IAt
 
         private SymbolCompletionState _state;
         private CustomAttributesBag<CSharpAttributeData> _lazyCustomAttributesBag;
+
+        /// <summary>
+        /// May hold the following values:
+        /// - <see cref="TypeParameterBounds.Unset"/>
+        /// - <see cref="TypeParameterBounds.NullFromLightweightBinding"/>
+        /// - null (indicates that we did full binding, but have no bounds)
+        /// - an actual value, which may be from lightweight or full binding, depending on <see cref="TypeParameterBounds.UsedLightweightTypeConstraintBinding"/>
+        ///
+        /// The distinction between <see cref="TypeParameterBounds.NullFromLightweightBinding"/> and null
+        /// helps us ensure that we do full binding at some point and thus produce diagnostics.
```

Comment:
What is an example where we need the distinction? (When would be have `null` from lightweight binding and produce diagnostics from full binding?)

Reply:
One scenario where the difference was visible was `where T : Span<int>`. The current diagnostic complains about `Span<int>` (and that constraint gets filtered out), but if we report the diagnostic from lightweight binding we'll complain about `Span<T>`.
It's not a big difference, but also we may add more such scenarios, so I'm not sure we want to chance it. Also, if we decided to keep the previous design, we'd want to add comments, as this was surprising when I had to step through.